fish-rust/src/env/environment.rs

```diff
@@ -371,8 +379,315 @@ lazy_static! {
     static ref PRINCIPAL_STACK: EnvStackRef = Arc::new(EnvStack::new());
 }
 
-// Note: this is an incomplete port of env_init(); the rest remains in C++.
-pub fn env_init(do_uvars: bool) {
+/// Some configuration path environment variables.
+const FISH_DATADIR_VAR: &wstr = L!("__fish_data_dir");
+const FISH_SYSCONFDIR_VAR: &wstr = L!("__fish_sysconf_dir");
+const FISH_HELPDIR_VAR: &wstr = L!("__fish_help_dir");
+const FISH_BIN_DIR: &wstr = L!("__fish_bin_dir");
+const FISH_CONFIG_DIR: &wstr = L!("__fish_config_dir");
+const FISH_USER_DATA_DIR: &wstr = L!("__fish_user_data_dir");
+
+/// Maximum length of hostname. Longer hostnames are truncated.
+const HOSTNAME_LEN: usize = 255;
+
+/// Function to get an identifier based on the hostname.
+fn get_hostname_identifier() -> Option<WString> {
+    // The behavior of gethostname if the buffer size is insufficient differs by implementation and
+    // libc version Work around this by using a "guaranteed" sufficient buffer size then truncating
+    // the result.
+    let mut b = [0 as libc::c_char; HOSTNAME_LEN + 1];
+    if unsafe { libc::gethostname(b.as_mut_ptr(), b.len()) } == 0 {
+        let cstr = unsafe { CStr::from_ptr(b.as_ptr()) };
+        let res = str2wcstring(cstr.to_bytes());
+
+        if res.is_empty() {
+            None
+        } else {
+            Some(res)
+        }
+    } else {
+        None
+    }
+}
+
+fn setup_user(vars: &EnvStack) {
+    let uid = unsafe { libc::geteuid() };
+    let user_var = vars.get_unless_empty(L!("USER"));
+
+    let mut userinfo: MaybeUninit<libc::passwd> = MaybeUninit::zeroed();
```

Comment:
interesting. I've been using `unsafe { std::mem::zeroed() }`, we should make things consistent in future.
MaybeUninit seems superior because it has `.as_mut_ptr()` etc.
Any particular reason not to use `MaybeUninit::uninit()`? That might allow tools to find bugs.

Reply:
`struct passwd userinfo;` was not zero initialized, it was uninitialized.

Reply:
If it is not already clear: I am not very familiar with C++

In that case I will change it to `uninit`